src/api/api.js

```diff
@@ -36,78 +35,85 @@ api_schema.register_api(require('./func_api'));
-            if (name === 'rpc' || name === 'options') throw new Error('ILLEGAL API ID');
-            this[name] = {};
-            _.each(api.methods, (method_api, method_name) => {
-                this[name][method_name] = (params, options) => {
-                    options = _.create(this.options, options);
-                    return rpc._request(api, method_api, params, options);
-                };
-            });
-        });
-    }
+const shared_client_proto = {
+    RPC_BUFFERS,
 
-    /**
-     * extend the rpc client prototype with convinient methods
-     */
-    create_auth_token(params) {
-        return this.auth.create_auth(params)
-            .then(res => {
-                this.options.auth_token = res.token;
-                return res;
-            });
-    }
+    create_auth_token: async function(params) {
+        const res = this.auth.create_auth(params);
+        this.options.auth_token = res.token;
+        return res;
+    },
 
-    create_access_key_auth(params) {
-        return this.auth.create_access_key_auth(params)
-            .then(res => {
-                this.options.auth_token = res.token;
-                return res;
-            });
-    }
+    create_access_key_auth: async function(params) {
+        const res = this.auth.create_access_key_auth(params);
+        this.options.auth_token = res.token;
+        return res;
+    },
 
-    async create_k8s_auth(params) {
+    create_k8s_auth: async function(params) {
         const res = await this.auth.create_k8s_auth(params);
         this.options.auth_token = res.token;
         return res;
+    },
+
+    _invoke_api(api, method_api, params, options) {
+        options = Object.assign(
+            Object.create(this.options),
+            options
+        );
+        return this.rpc._request(api, method_api, params, options);
     }
+};
 
+function new_client_factory(rpc) {
+    // Copy the shared part instead of using it as a proto
+    // to skip one hoop when invoking these methods.
+    const client_proto = {
+        ...shared_client_proto,
+        rpc
+    };
+
+    for (const api of Object.values(rpc.schema)) {
+        if (!api || !api.id || api.id[0] === '_') {
+             continue;
+        }
+
+        // Skip common api and other apis that do not defien methods.
+        if (!api.methods) {
+            continue;
+        }
+
+        const name = api.id.replace(/_api$/, '');
+        if (name === 'rpc' || name === 'options') {
+            throw new Error('ILLEGAL API ID');
+        }
+
+        const api_proto = {};
+        for (const [method_name, method_api] of Object.entries(api.methods)) {
+            // The following getter is defined as a function and not as an arrow function
+            // to prevent the capture of the surrounding context when using this
+            api_proto[method_name] = function(params, options) {
+                return this.client._invoke_api(api, method_api, params, options);
```

Comment:
Can you add a comment that explains what is `this` in this context?
I think that our linter also requires it for non-class functions that use this by adding a jsdoc comment `/** @this {TYPE} */`

Reply:
I will, no problem. 
FYI, the linter will only complain if it was an arrow function. Making this a traditional function fix all of these problem

Reply:
done

Reply:
This is the rule - https://eslint.org/docs/rules/no-invalid-this
Arrow functions are not relevant to it at all as this in them will resolve to outer context.
But for normal functions this rule makes sure `this` is used only in "class" context.